File: src/application/queries/ticket_health_handler.py
"""Application query handler for ticket health reporting.

TicketHealthHandler reads open tickets via a TicketReaderProtocol
and builds a TicketHealthReport covering freshness flags, staleness,
and review status.
"""

from __future__ import annotations

from typing import Protocol, runtime_checkable

from src.domain.models.ticket_freshness import (
    FlaggedTicket,
    FreshnessFlag,
    OpenTicketData,
    TicketFreshnessStatus,
    TicketHealthReport,
)


@runtime_checkable
class TicketReaderProtocol(Protocol):
    """Port interface for reading ticket data from the issue tracker."""

    def read_open_tickets(self) -> tuple[OpenTicketData, ...]: ...

    def read_flags(self, ticket_id: str) -> tuple[FreshnessFlag, ...]: ...


class TicketHealthHandler:
    """Query handler that builds a TicketHealthReport from the backlog.

    Reads all open tickets, identifies those needing review, gathers
    their freshness flags, and computes the oldest last-reviewed date.
    """

    def __init__(self, reader: TicketReaderProtocol) -> None:
        self._reader = reader

    def report(self) -> TicketHealthReport:
        """Build a ticket health report.

        Returns:
            TicketHealthReport with flagged tickets, totals, and staleness info.
        """
        open_tickets = self._reader.read_open_tickets()

        flagged: list[FlaggedTicket] = []
        for ticket in open_tickets:
            if "review_needed" in ticket.labels:
                flags = self._reader.read_flags(ticket.ticket_id)
                flagged.append(
                    FlaggedTicket(
                        ticket_id=ticket.ticket_id,
                        title=ticket.title,
                        flags=flags,
                        status=TicketFreshnessStatus.REVIEW_NEEDED,
                    )
                )

        # Find the oldest last_reviewed across ALL open tickets
        reviewed_dates = [t.last_reviewed for t in open_tickets if t.last_reviewed is not None]
        oldest = min(reviewed_dates) if reviewed_dates else None

        return TicketHealthReport(
            flagged_tickets=tuple(flagged),
            total_open=len(open_tickets),
            oldest_last_reviewed=oldest,
        )

File: internal/ticket/application/ticket_health_handler.go
package application

import (
	"context"
	"fmt"

	"github.com/alty-cli/alty/internal/ticket/domain"
)

// TicketReader is a handler-local interface for reading ticket data.
// Defined where consumed per Go convention.
type TicketReader interface {
	// ReadOpenTickets returns all open tickets.
	ReadOpenTickets(ctx context.Context) ([]domain.OpenTicketData, error)

	// ReadFlags returns freshness flags for a specific ticket.
	ReadFlags(ctx context.Context, ticketID string) ([]domain.FreshnessFlag, error)
}

// TicketHealthHandler queries ticket freshness and builds a health report.
// It reads all open tickets, identifies those needing review, gathers
// their freshness flags, and computes the oldest last-reviewed date.
type TicketHealthHandler struct {
	reader TicketReader
}

// NewTicketHealthHandler creates a new TicketHealthHandler.
func NewTicketHealthHandler(reader TicketReader) *TicketHealthHandler {
	return &TicketHealthHandler{reader: reader}
}

// Report builds a ticket health report.
func (h *TicketHealthHandler) Report(ctx context.Context) (domain.TicketHealthReport, error) {
	openTickets, err := h.reader.ReadOpenTickets(ctx)
	if err != nil {
		return domain.TicketHealthReport{}, fmt.Errorf("reading open tickets: %w", err)
	}

	var flagged []domain.FlaggedTicket
	for _, ticket := range openTickets {
		if hasLabel(ticket.Labels(), "review_needed") {
			flags, err := h.reader.ReadFlags(ctx, ticket.TicketID())
			if err != nil {
				return domain.TicketHealthReport{}, fmt.Errorf("reading flags for ticket %s: %w", ticket.TicketID(), err)
			}
			flagged = append(flagged, domain.NewFlaggedTicket(
				ticket.TicketID(),
				ticket.Title(),
				flags,
				domain.FreshnessStatusReviewNeeded,
			))
		}
	}

	// Find the oldest last_reviewed across ALL open tickets.
	var oldest *string
	for _, ticket := range openTickets {
		lr := ticket.LastReviewed()
		if lr != nil {
			if oldest == nil || *lr < *oldest {
				s := *lr
				oldest = &s
			}
		}
	}

	return domain.NewTicketHealthReport(flagged, len(openTickets), oldest), nil
}

// hasLabel checks if a label is present in a slice.
func hasLabel(labels []string, target string) bool {
	for _, l := range labels {
		if l == target {
			return true
		}
	}
	return false
}
